lib/reader-watch.js

```diff
@@ -0,0 +1,133 @@
+import { once } from 'events'
+
+import fs from 'node:fs'
+import fsPromises from 'node:fs/promises'
+
+import Reader from './reader.js'
+import { ENOENT, isFileNotThereError } from './utils/errors.js'
+import { noop } from './utils/misc.js'
+
+/** @implements {Pick<Reader, keyof Reader>} */
+export default class ReaderWatch {
```

Comment:
would be cool if `ReaderWatch` extended `Reader` but maybe that's dismissing the potential implementation complexity 😄 the scattered `Pick<Reader, keyof Reader>` feel a little strange at initial glance

Reply:
It implements reader, but it can't extend it because it replaces the reader instance when the file changes. Unfortunately to "implement" a class from TS perspective all the private methods also need to match. The Pick trick says "it's the same, but only the public methods"

Reply:
Yeah it makes sense upon reading the implementation.

> Unfortunately to "implement" a class from TS perspective all the private methods also need to match

Kind of surprised by this. Then again, haven't worked heavily with TS in that specific way